packages/ref-celery/src/cmip_ref_celery/executor.py

```diff
@@ -0,0 +1,126 @@
+
+    We cannot resume tasks that are in progress if the process terminates.
+    That should be possible tracking some additional state in the database.
+    """
+
+    name = "celery"
+
+    def __init__(self):
+        self._results: list[celery.result.AsyncResult] = []
+
+    def run_metric(
+        self,
+        provider: MetricsProvider,
+        metric: Metric,
+        definition: MetricExecutionDefinition,
+        metric_execution_result: MetricExecutionResult | None = None,
+    ) -> None:
+        """
+        Run a metric calculation asynchronously
+
+        This will queue the metric to be run by a Celery worker.
+        The results will be stored in the database when the task completes if `metric_execution_result`
+        is specified.
+        No result will be returned from this function.
+        Instead, you can periodically check the status of the task in the database.
+
+        Parameters
+        ----------
+        provider
+            Provider for the metric
+        metric
+            Metric to run
+        definition
+            A description of the information needed for this execution of the metric
+            This include relative paths to the data files,
+            which will be converted to absolute paths when being executed
+        metric_execution_result
+            Result of the metric execution
+            This is a database object that contains the results of the execution.
+            If provided, it will be updated with the results of the execution.
+            This may happen asynchronously, so the results may not be immediately available.
+        """
+        from cmip_ref_celery.worker_tasks import handle_result
+
+        name = generate_task_name(provider, metric)
+
+        async_result = app.send_task(
+            name,
+            args=[
+                definition,
+            ],
+            queue=provider.slug,
+            link=handle_result.s(metric_execution_result_id=metric_execution_result.id).set(queue="celery")
+            if metric_execution_result
+            else None,
+        )
+        logger.debug(f"Celery task {async_result.id} submitted")
+        self._results.append(async_result)
+
+    def join(self, timeout: float) -> None:
+        """
+        Wait for all executions to finish
+
+        This will block until all executions have finished running,
+        and display a progress bar while waiting.
+
+        Any tasks still running after the timeout will continue to run in the background.
+
+        Parameters
+        ----------
+        timeout
+            Maximum time to wait in seconds before raising a TimeoutError
+
+        Raises
+        ------
+        TimeoutError
+            If all executions aren't completed within the specified timeout
+        """
+        start_time = time.time()
+        refresh_time = 0.5  # Time to wait between checking for completed tasks in seconds
+
+        results = self._results.copy()
```

Comment:
I don't know about the expected usage pattern, but if you call `run_metric` repeatedly, and every once in a while `join`, you'll leak ready results objects in `self._results`. For a steady-state thing where we run new metrics whenever needed and `join` every once in a while, the memory usage will grow without bounds. Is there a reason for this copy here? If you just don't call `.copy()` here, the leak goes away.

Reply:
Good pickup. I intended to remove completed results from `self._results` upon completion. This is deliberately opaque to which results have been successfully as I don't want any consumers to rely on completed results until we have a proper solution